modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/FilePerformanceStatisticsReader.java

```diff
@@ -0,0 +1,342 @@
+import org.apache.ignite.internal.util.GridUnsafe;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.lang.IgniteUuid;
+import org.jetbrains.annotations.Nullable;
+
+import static java.nio.ByteBuffer.allocateDirect;
+import static java.nio.ByteOrder.nativeOrder;
+import static java.nio.file.Files.walkFileTree;
+import static org.apache.ignite.internal.processors.performancestatistics.OperationType.JOB;
+import static org.apache.ignite.internal.processors.performancestatistics.OperationType.QUERY;
+import static org.apache.ignite.internal.processors.performancestatistics.OperationType.QUERY_READS;
+import static org.apache.ignite.internal.processors.performancestatistics.OperationType.TASK;
+import static org.apache.ignite.internal.processors.performancestatistics.OperationType.TX_COMMIT;
+import static org.apache.ignite.internal.processors.performancestatistics.OperationType.cacheOperation;
+import static org.apache.ignite.internal.processors.performancestatistics.OperationType.transactionOperation;
+
+/**
+ * Walker over the performance statistics file.
+ *
+ * @see FilePerformanceStatisticsWriter
+ */
+public class FilePerformanceStatisticsReader {
+    /** File read buffer size. */
+    private static final int READ_BUFFER_SIZE = (int)(8 * U.MB);
+
+    /** Uuid as string pattern. */
+    private static final String UUID_STR_PATTERN =
+        "[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}";
+
+    /** File name pattern. */
+    private static final Pattern FILE_PATTERN = Pattern.compile("^node-(" + UUID_STR_PATTERN + ").prf$");
+
+    /** IO factory. */
+    private static final RandomAccessFileIOFactory ioFactory = new RandomAccessFileIOFactory();
+
+    /**
+     * Walks over performance statistics files.
+     *
+     * @param filesOrDirs Files or directories.
+     * @param handlers Handlers to process deserialized operation.
+     */
+    public static void read(List<File> filesOrDirs, PerformanceStatisticsHandler... handlers) throws IOException {
+        List<File> files = resolveFiles(filesOrDirs);
+
+        if (files.isEmpty())
+            return;
+
+        for (File file : files)
+            readFile(file, handlers);
+    }
+
+    /**
+     * Walks over performance statistics file.
+     *
+     * @param file Performance statistics file.
+     * @param handlers Handlers to process deserialized operation.
+     */
+    private static void readFile(File file, PerformanceStatisticsHandler... handlers) throws IOException {
+        UUID nodeId = checkFileName(file);
+
+        ByteBuffer buf = allocateDirect(READ_BUFFER_SIZE).order(nativeOrder());
+
+        PerformanceStatisticsDeserializer des = new PerformanceStatisticsDeserializer(nodeId, handlers);
+
+        try (FileIO io = ioFactory.create(file)) {
+            while (true) {
+                int read = io.read(buf);
+
+                buf.flip();
+
+                if (read <= 0)
```

Comment:
Nit: We should firstly check read results and after flip the buffer.

Reply:
Done.